docs/man/rpm-macros.7.scd

```diff
@@ -0,0 +1,760 @@
+
+*%{getenv:*_NAME_*}*
+	*getenv*(3) macro analogue.
+
+	Example:
+	```
+	%{getenv:HOME}
+	```
+
+*%rpmversion*
+	Expand to running rpm version.
+
+## Output
+*%{echo:*_STRING_*}*
+	Print _STRING_ to process stdout.
+
+	Example:
+	```
+	%{echo:Building with foo}
+	```
+
+*%{warn:*_STRING_*}*
+	Print _STRING_ prefixed with "warning: " to process stderr.
+
+	Example:
+	```
+	%{warning:Foo is deprecated}
+	```
+
+*%{error:*_STRING_*}*
+	Print _STRING_ prefixed with "error: " to process stderr and
+	flag an error in the macro processor.
+
+	Example:
+	```
+	%{error:Invalid argument}
+	```
+
+*%verbose*
+	Expand to 1/0 whether RPM is in verbose mode or not. (Added: 4.17.1)
+
+*%{verbose:*_STRING_*}*
+	Expand to _STRING_ if rpm is in verbose mode, the empty string if not.
+	(Added: 4.17.1)
+
+	Example:
+	```
+	%{verbose:-x}
+	```
+
+## Spec specific macros
+*%{S:*_NUMBER_*}*
+	Expand to the file name of the specified Source _NUMBER_.
+	*%{S:n}* is equivalent to *%{SOURCEn}*.
+
+*%{P:*_NUMBER_*}*
+	Expand to the file name of the specified Patch _NUMBER_.
+	*%{P:n}* is equivalent to *%{PATCHn}*.
+
+## Diagnostics
+*%trace*
+	Toggle print of debugging information before/after expansion.
+
+*%dump*
+	Print the active (i.e. non-covered) macro table.
+
+*%\_\_file_name*
+	Expand to current file name (if parsing a file). (Added: 4.15)
+
+*%\_\_file_lineno*
+	Expand to current line number in current file (if parsing a file).
+	(Added: 4.15)
+
+# EXAMPLES
+## Example 1. Define a simple macro
+Define macro *mylib* to a path relative to *%{\_libdir}* macro in a spec:
+
+```
+%define mylib %{_libdir}/mine
+```
+
+## Example 2. Define a parametric macro
+Define parametric macro *myhelper* which executes the program
+specified by *%myprog* with it's first argument and always passing
+the option *--some-opt* to it, and additionally the *--xtra* option
+if it received the *-x* option itself:
+
+```
+%define myhelper(x) %{myprog} --some-opt %{?-x:--xtra} %{1}
+```
+
+## Example 3. Define a macro utilizing shell expansion
+Define macro *%today* that expands to the current date in _YYMMDD_
+format by calling the *date*(1) shell utility.
+Note the 2nd *%* needed to escape the arguments to *date*(1):
+
+```
+%define today %(date +%%y%%m%%d)
+```
+
+## Example 4. Define a macro conditionally
+Define macro *mypath* if it wasn't previously defined:
+
+```
+%{!?mypath: %define mypath /some/where}
+```
+
+## Example 5. Conditional expansion
+Expands to *1* if *use_foo* is defined and *0* otherwise:
+
+```
+%{?use_foo:1}%{!?use_foo:0}
+```
+
+## Example 6. Expressions
+Calculate 5 \* 1024:
+
+```
+%[5 * 1024]
+```
+
+Expand to literal *true* or *false* depending on a condition:
+
+```
+%[1 < 2 ? "true" : "false"]
+```
+
+Compare versions, expanding to *1* or *0* on true/false:
+
+```
+%[ v"3.1.0-1" < v"1.0~alpha-2" ? 1 : 0]
+```
+
+Expands to *1* if *%aa* expands to *5*, otherwise expands to *2*:
+
+```
+%[ "%{aa}" == "5" ? 1 : 2]
+```
+
+# DEBUGGING
```

Comment:
I'd add an example for `%{macrobody: ... }`, too. It's something I've only learned today (by reading about it here) and immediately thought - yes, that's what I needed many times in the past when debugging some unexpected macro expansions! :smile: